llvm/lib/Target/RISCV/RISCVFrameLowering.cpp

```diff
@@ -117,6 +117,39 @@ getRestoreLibCallName(const MachineFunction &MF,
   return RestoreLibCalls[LibCallID];
 }
 
+static bool shouldSignReturnAddress(const MachineFunction &MF) {
+  // The function should be signed in the following situations:
+  // - sign-return-address=all and RV32
+  // - sign-return-address=non-leaf and the functions spills the
+  //   X1(return address) and RV32
+  // - sign-return-address=non-leaf and
+  //   MF.getFrameInfo().isCalleeSavedInfoValid()==false and RV32
+
+  const Function &F = MF.getFunction();
+  if (!F.hasFnAttribute("sign-return-address"))
+    return false;
+
+  StringRef Scope = F.getFnAttribute("sign-return-address").getValueAsString();
+  if (Scope.equals("none"))
+      return false;
+
+  // Only 32bit RISCV is supported
+  auto &Subtarget = MF.getSubtarget<RISCVSubtarget>();
+  if (Subtarget.getTargetABI() != RISCVABI::ABI_ILP32)
+    llvm_unreachable("Unsupported ABI");
+
+  if (Scope.equals("all"))
+    return true;
+
+  assert(Scope.equals("non-leaf") && "Expected all, none or non-leaf");
```

Comment:
It would be nice if the attribute parsing logic was all in one place, in a public function. Feel free to add some kind of public method for it, maybe on RISCVSubtarget? You could return an enum for All/None/Non-Leaf.

(I realise some of the logic in this function is specifically about detecting whether a function is non-leaf, that should probably stay below).

Reply:
@lenary  Thank you for your feedback. I put the attribute parsing logic together. Please confirm.